package/samplers/grey_wolf_optimization/grey_wolf_optimization.py

```diff
@@ -0,0 +1,103 @@
+        self._rng = np.random.RandomState(seed)
+        self.dim = 0
+        self.leaders: np.ndarray = np.array([])  # Leaders (alpha, beta, gamma, ...) positions
+        self.wolves: np.ndarray = np.array([])  # Wolf positions
+        self.fitnesses: np.ndarray = np.full(population_size, np.inf)  # Fitness values
+        self._random_sampler = RandomSampler(seed=seed)
+        self.queue: list[dict[str, Any]] = []  # Queue to hold candidate positions
+
+    def _lazy_init(self, search_space: dict[str, BaseDistribution]) -> None:
+        self.dim = len(search_space)
+        self.lower_bound = np.array([dist.low for dist in search_space.values()])
+        self.upper_bound = np.array([dist.high for dist in search_space.values()])
+        self.wolves = (
+            np.random.rand(self.population_size, self.dim) * (self.upper_bound - self.lower_bound)
+            + self.lower_bound
+        )
+        self.leaders = np.zeros((self.num_leaders, self.dim))  # Initialize as zeros
+
+    def sample_relative(
+        self,
+        study: optuna.Study,
+        trial: optuna.FrozenTrial,
+        search_space: dict[str, BaseDistribution],
+    ) -> dict[str, Any]:
+        if len(search_space) == 0:
+            return {}
+
+        if self.dim != len(search_space):
+            self._lazy_init(search_space)
+
+        if len(self.queue) != 0:
+            return self.queue.pop(0)
+
+        if len(study.trials) < self.population_size:
+            # Fill the initial population using sample_independent
+            new_position = {
+                k: self.sample_independent(study, trial, k, dist)
+                for k, dist in search_space.items()
+            }
+            self.wolves[len(study.trials)] = np.array(list(new_position.values()))
+            return new_position
+
+        if len(study.trials) % self.population_size == 0:
+            # Perform one iteration of GWO
+            completed_trials = study.get_trials(states=[optuna.trial.TrialState.COMPLETE])
+            self.fitnesses = np.array(
+                [trial.value for trial in completed_trials[-self.population_size :]]
+            )
+
+            # Update leaders (alpha, beta, gamma, ...)
+            sorted_indices = np.argsort(self.fitnesses)
+            self.leaders = self.wolves[sorted_indices[: self.num_leaders]]
+
+            # Linearly decrease from 2 to 0
+            a = 2 * (1 - len(study.trials) / (self.max_iter * self.population_size))
+
+            # Calculate A, C, D, X values for position update
+            r1 = self._rng.rand(self.population_size, self.num_leaders, self.dim)
+            r2 = self._rng.rand(self.population_size, self.num_leaders, self.dim)
+            A = 2 * a * r1 - a
+            C = 2 * r2
+            D = np.abs(C * self.leaders - self.wolves[:, np.newaxis, :])
+            X = self.leaders - A * D
+
+            # Update wolves' positions and store them in the queue
+            self.wolves = np.mean(X, axis=1)
```

Comment:
We may want to clip the values in case they exceed the range. But we can work on it in a follow-up PR.